frontend/stmt_inferrer.py

```diff
@@ -453,18 +457,64 @@ def _infer_class_def(node, context, solver):
+            # If the attribute is a non static method, set the type of the first arg to be an instance of this class
+            args_len = class_to_funcs[attr][0]
+            arg_accessor = getattr(solver.z3_types.type_sort, "func_{}_arg_1".format(args_len))
+            solver.add(arg_accessor(class_attrs[attr]) == instance_type,
+                       fail_message="First arg in instance methods has class instance type")
+
+        if attr != "__init__" and attr in base_class_to_funcs:
+            # attr is an overridden method
+            # class_to_funcs[attr] is a tuple of two elements.
+            # The first is the args length, the second is the decorators list, the third is default args length
+            base_args_len = base_class_to_funcs[attr][0]
+            base_defaults_len = base_class_to_funcs[attr][2]
+            base_non_defaults_len = base_args_len - base_defaults_len
+            sub_args_len = class_to_funcs[attr][0]
+            sub_defaults_len = class_to_funcs[attr][2]
+            sub_non_defaults_len = sub_args_len - sub_defaults_len
+            decorators = base_class_to_funcs[attr][1] + class_to_funcs[attr][1]
+
+            if "staticmethod" in decorators:
+                if "staticmethod" not in base_class_to_funcs[attr][1]:
+                    raise TypeError("Static method {} in class {} cannot override non-static method.".format(attr,
+                                                                                                             node.name))
+                if "staticmethod" not in class_to_funcs[attr][1]:
+                    raise TypeError("Non-static method {} in class {} cannot override static method.".format(attr,
+                                                                                                             node.name))
+            else:
+                if not base_args_len:
+                    raise TypeError("Instance methods should have at least one argument (the receiver)."
+                                    "If you wish to create a static method, please add the appropriate decorator.")
+                if base_args_len > sub_args_len:
+                    raise TypeError("Method {} in subclass {} should have total arguments length more than or equal "
+                                    "that in superclass.".format(attr, node.name))
+                if base_non_defaults_len > sub_non_defaults_len:
+                    raise TypeError("Method {} in subclass {} should have non-default arguments length less than"
+                                    "or equal that in superclass".format(attr, node.name))
+
+                # handle arguments and return contravariance/covariance
+                for i in range(1, base_args_len):
+                    arg_accessor = getattr(solver.z3_types.type_sort, "func_{}_arg_{}".format(base_args_len, i + 1))
+                    solver.add(solver.z3_types.subtype(arg_accessor(base_attrs[attr]),
```

Comment:
Hm yeah, the soft constraints resulting from the actual code are probably more important, I agree.